spark/src/main/scala/org/apache/spark/sql/delta/actions/TableFeatureSupport.scala

```diff
@@ -281,12 +281,34 @@ trait TableFeatureSupport { this: Protocol =>
     (this.readerAndWriterFeatureNames -- to.readerAndWriterFeatureNames).size == 1
   }
 
+  /**
+   * This is a special protocol version downgrade path that can be invoked by setting legacy
+   * protocol versions on a table features protocol. This is the only downgrade path that may
+   * occur outside the DROP FEATURE command. The main requirement is the table features protocol
+   * contains only legacy features.
+   *
+   * During downgrade, we first add the relevant implicit legacy features to the protocol
+   * and then call downgradeProtocolVersionsIfNeeded. At this stage, the validation stage,
+   * we only have available the the start and end states of the Protocol, i.e. the intermediate
+   * state with all the required implicit features is not available. It is
+   * downgradeProtocolToLegacyVersionsIfNeeded which validates whether the intermediate state has
+   * all required legacy features are in the protocol before downgrading the protocol versions.
+   */
+  def canDowngradeToLegacy(to: Protocol): Boolean = {
+    require(supportsWriterFeatures)
+    require(!to.supportsWriterFeatures)
+
+    readerAndWriterFeatures.forall(_.isLegacyFeature) &&
```

Comment:
Now that I read this, I realize that it's actually an upgrade in some sense. :) It would be good to sprinkle that idea throughout the code in comments, so that people realize that it's not a real downgrade. 

```suggestion
    // Note that the downgrade can *add* legacy features. It's technically
    // a protocol *upgrade* then in terms of features but a *downgrade*
    // in terms of protocol versions.
    readerAndWriterFeatures.forall(_.isLegacyFeature) &&
```